contracts/src/chainlink/solidity/emergency/StarkNetValidator.sol

```diff
@@ -74,15 +120,137 @@ contract StarkNetValidator is TypeAndVersionInterface, AggregatorValidatorInterf
+    STARKNET_CROSS_DOMAIN_MESSENGER.sendMessageToL2{value: fee}(
+      L2_UPTIME_FEED_ADDR,
+      SELECTOR_STARK_UPDATE_STATUS,
+      payload
+    );
     return true;
   }
+
+  /// @notice L1 oracle is asked for a fast L1 gas price, and the price multiplied by the configured gas estimate
+  function _approximateFee() internal view returns (uint256) {
+    (, int256 fastGasPriceInWei, , , ) = AggregatorV3Interface(s_gasConfig.gasPriceL1Feed).latestRoundData();
+    return (uint256(fastGasPriceInWei) * s_gasConfig.gasEstimate);
+  }
+
+  /**
+   * @notice The selector is the starknet_keccak hash of the function name
+   * @dev StarkNet keccak is defined as the first 250 bits of the Keccak256 hash.
+   *   This is just Keccak256 augmented in order to fit into a field element.
+   * @param fn string function name
+   */
+  function _selectorStarkNet(string memory fn) internal pure returns (uint256) {
+    bytes32 digest = keccak256(abi.encodePacked(fn));
+    return uint256(digest) % 2**250; // get last 250 bits
+  }
+
+  function setGasConfig(uint256 gasEstimate, address gasPriceL1Feed) external onlyOwnerOrConfigAccess {
+    _setGasConfig(gasEstimate, gasPriceL1Feed);
+    emit GasConfigSet(gasEstimate, gasPriceL1Feed);
+  }
+
+  /// @notice internal method that stores the gas configuration
+  function _setGasConfig(uint256 gasEstimate, address gasPriceL1Feed) internal {
+    if (gasPriceL1Feed == address(0)) {
+      revert InvalidGasPriceL1FeedAddress();
+    }
+    s_gasConfig = GasConfig(gasEstimate, gasPriceL1Feed);
+    emit GasConfigSet(gasEstimate, gasPriceL1Feed);
+  }
+
+  /**
+   * @notice sets config AccessControllerInterface contract
+   * @dev only owner can call this
+   * @param accessController new AccessControllerInterface contract address
+   */
+  function setConfigAC(address accessController) external onlyOwner {
+    _setConfigAC(accessController);
+  }
+
+  /// @notice Internal method that stores the configuration access controller
+  function _setConfigAC(address accessController) internal {
+    address previousAccessController = address(s_configAC);
+    if (accessController != previousAccessController) {
+      s_configAC = AccessControllerInterface(accessController);
+      emit ConfigACSet(previousAccessController, accessController);
+    }
+  }
+
+  /**
+   * @notice sets the source aggregator AggregatorInterface contract
+   * @dev only owner can call this
+   * @param source the source aggregator that we'll read data from (on retries)
+   */
+  function setSourceAggregator(address source) external onlyOwner {
+    _setSourceAggregator(source);
+  }
+
+  /// @notice Internal method that sets the source aggregator AggregatorInterface contract
+  function _setSourceAggregator(address source) internal {
+    address previousSource = address(s_source);
+    if (source != previousSource) {
+      s_source = AggregatorV3Interface(source);
```

Comment:
I removed the part about adding and removing access as the aggregator never writes to the Validator so it doesn't need any access.  Calling `this.addAccess` and `this.removeAccess` also does not work when we try to initialize the aggregator in the constructor.

Reply:
Please add the in-line docs back, somewhere maybe here:

```
// NOTICE: we don't give access to the new source aggregator
// It is not always the case that the source aggregator is also the sender for the 'validate' invocation
// as we usually deploy an additional proxy in between (owner can give access):
//   https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/ValidatorProxy.sol
```

It's helpful to document these things for anyone reading the code, and us once we'll need to go back and work on access control improvements for these contracts.